src/modules/ekf2/EKF/mag_3d_control.cpp

```diff
@@ -0,0 +1,215 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2023 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file mag_3d_control.cpp
+ * Control functions for ekf mag 3D fusion
+ */
+
+#include "ekf.h"
+
+void Ekf::controlMag3DFusion(const magSample &mag_sample, const bool common_starting_conditions_passing,
+			     estimator_aid_source3d_s &aid_src)
+{
+	static constexpr const char *AID_SRC_NAME = "mag";
+
+	resetEstimatorAidStatus(aid_src);
+
+	const bool wmm_updated = (_wmm_gps_time_last_set > aid_src.time_last_fuse);
```

Comment:
This is a bit misleading: the wmm is constantly updated, but we only need to do a reset when the wmm is set for the first time.

Reply:
Actually it's only updated twice, first when we get a 2d fix with eph < 1000, then again when the NED origin is initialised.

https://github.com/PX4/PX4-Autopilot/blob/444e5d2d4abbb0adaabf1ad3cddaa2f9d2f9ee2b/src/modules/ekf2/EKF/gps_checks.cpp#L63C30-L63C53

It would probably be better to continue updating it periodically just in case anyone wants to fly around the world.

Reply:
Right, we should update it periodically (but only trigger a reset after 2d fix and NED init)

Reply:
I changed it to check for WMM changes periodically, but only update if the declination or inclination changed by more than a degree (better arbitrary threshold?).